### PR TITLE
fix overlay odds

### PR DESCRIPTION
**Pull request by Spookerton**

### CODE REVIEW

**Comment by Spookerton** [comment]
@Atermonera I'm not sure why VSC is deciding to save out entire file encoding changes for me but it's probably the `"files.encoding": "utf8"` part of the workspace settings. In terms of outcome it's a good thing, but very annoying for review obviously, so I'll work out what and roll it back.

File: code/controllers/subsystems/overlays.dm [hunk 1]
```diff
@@ -1,242 +1,239 @@
...
-	flags = SS_TICKER
-	wait = 1 // SS_TICKER - Ticks
-	priority = FIRE_PRIORITY_OVERLAYS
-	init_order = INIT_ORDER_OVERLAY
-
-	/// The queue of atoms that need overlay updates.
-	var/static/tmp/list/queue = list()
-
-	/// An image used to create appearances to be cached.
-	var/static/tmp/image/renderer = new
```
Comment (Spookerton) [inline code/controllers/subsystems/overlays.dm:12]: removes this

File: code/controllers/subsystems/overlays.dm [hunk 2]
```diff
@@ -1,242 +1,239 @@
...
+
+
+/datum/controller/subsystem/overlays/Recover()
+	queue.Cut()
+	state_cache.Cut()
+	icon_cache.Cut()
+	cache_size = 0
+	for (var/atom/atom)
+		atom.flags &= ~OVERLAY_QUEUED
+		CHECK_TICK
```
Comment (Spookerton) [inline code/controllers/subsystems/overlays.dm:28]: adds this

File: code/controllers/subsystems/overlays.dm [hunk 3]
```diff
@@ -1,242 +1,239 @@
...
+	queue.Cut()
+
+
+/datum/controller/subsystem/overlays/proc/GetStateAppearance(icon, state)
+	var/list/subcache = state_cache[icon]
+	if (!subcache)
+		subcache = list()
+		state_cache[icon] = subcache
+	if (!subcache[state])
+		var/image/image = new (icon, null, state)
```
Comment (Spookerton) [inline code/controllers/subsystems/overlays.dm:58]: replaces renderer with a clean image here

File: code/controllers/subsystems/overlays.dm [hunk 4]
```diff
@@ -1,242 +1,239 @@
...
+		subcache[state] = image.appearance
+		++cache_size
+	return subcache[state]
+
+
+/datum/controller/subsystem/overlays/proc/GetIconAppearance(icon)
+	if (!icon_cache[icon])
+		var/image/image = new (icon)
```
Comment (Spookerton) [inline code/controllers/subsystems/overlays.dm:66]: replaces renderer with a clean image here

File: code/controllers/subsystems/overlays.dm [hunk 5]
```diff
@@ -1,242 +1,239 @@
...
+			result += GetIconAppearance(entry)
+		else
+			if (isloc(entry))
+				if (entry.flags & OVERLAY_QUEUED)
+					entry.ImmediateOverlayUpdate()
+			if (!ispath(entry))
+				result += entry.appearance
+			else
+				var/image/image = entry
+				result += image.appearance
```
Comment (Spookerton) [inline code/controllers/subsystems/overlays.dm:94]: replaces passing appearance through an image with just using it, still allows the questionable path behavior by giving it a clean image

File: code/game/objects/structures/crates_lockers/closets/secure/guncabinet.dm [hunk 6]
```diff
@@ -1,51 +1,50 @@
...
+	. = ..()
+	update_icon()
+
+/obj/structure/closet/secure_closet/guncabinet/toggle()
+	..()
+	update_icon()
+
+/obj/structure/closet/secure_closet/guncabinet/update_icon()
+	cut_overlays()
+	var/list/add = list()
```
Comment (Spookerton) [inline code/game/objects/structures/crates_lockers/closets/secure/guncabinet.dm:18]: replaces doing a bunch of individual adds and queues with building a list first and then passing it all at once

File: code/modules/hydroponics/trays/tray_update_icons.dm [hunk 7]
```diff
@@ -1,92 +1,92 @@
...
+			name = "[base_name] ([seed.seed_name])"
+		else
+			name = "[seed.seed_name]"
+	else
+		name = initial(name)
+
+	if(labelled)
+		name += " ([labelled])"
+
+	cut_overlays()
```
Comment (Spookerton) [inline code/modules/hydroponics/trays/tray_update_icons.dm:15]: corrects this

File: code/controllers/subsystems/overlays.dm (hunk 2 above)
Comment (Lohikar) [reply code/controllers/subsystems/overlays.dm:28]: Recover shouldn't sleep afaik, it's better to lag than risk something interacting with the subsystem while it's recovering.

**Review by Atermonera — APPROVED** [review]

**Pull request merged**